src/sync/patch.rs

```diff
@@ -0,0 +1,203 @@
+use crate::db;
+use nanoserde::DeJson;
+use std::default::Default;
+use str_macro::str;
+
+const OP_ADD: &str = "add";
+const OP_REMOVE: &str = "remove";
+const OP_REPLACE: &str = "replace";
+
+#[derive(Clone, Debug, Default, DeJson, PartialEq)]
+pub struct Operation {
+    pub op: String,
+    pub path: String,
+    #[nserde(rename = "valueString")]
+    #[nserde(default)]
+    pub value_string: String,
+}
+
+pub fn apply(db_write: &mut db::Write, patch: &[Operation]) -> Result<(), PatchError> {
+    use PatchError::*;
+    for op in patch.iter() {
+        if op.path[..1] != str!("/") {
+            return Err(InvalidPath(op.path.clone()));
+        }
+        // Strip first character from path.
+        let mut chars = op.path.chars();
+        chars.next();
+        let key = chars.as_str().as_bytes().to_vec();
+
+        match op.op.as_str() {
+            OP_ADD | OP_REPLACE => {
+                let value = op.value_string.as_bytes().to_vec();
+                db_write.put(key, value);
+            }
+            // Should we error if we try to remove a key that doesn't exist?
+            OP_REMOVE => {
+                if key.is_empty() {
+                    // Top-level remove (path == "/").
+                    db_write.reset_map();
+                } else {
+                    db_write.del(key);
+                }
+            }
+            _ => return Err(InvalidOp(op.op.to_string())),
+        };
+    }
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum PatchError {
+    InvalidOp(String),
+    InvalidPath(String),
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::dag;
+    use crate::db;
+    use crate::db::test_helpers::*;
+    use crate::kv::memstore::MemStore;
+    use std::collections::HashMap;
+
+    macro_rules! map(
```

Comment:
`hashmap!` is already checked in: https://github.com/rocicorp/repc/blob/master/src/util/nanoserde/any.rs#L126

Reply:
thanks will use that next time!